src/main/java/org/opensearch/security/privileges/CheckTable.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+package org.opensearch.security.privileges;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * A specialized data structure which represents a matrix of checkmarks. During construction, you can specify its rows
+ * and columns. Initially, all cells of the matrix are unchecked. You can use the check() method to set particular
+ * cells to checked(). There are various methods to query the state of the whole matrix, of columns and rows.
+ *
+ * The state of the matrix can be visualized with the toTableString() method:
+ *
+ * <pre>
+ *          | indices:data/read/search |
+ * index_a11| ok                       |
+ * index_a12| MISSING                  |
+ * index_a13| MISSING                  |
+ * index_a14| MISSING                  |
+ * </pre>
+ *
+ * TODO: This is for now just a minimalistic, inefficient and non-production-ready implementation. A better replacement is
+ * necessary.
+ */
```

Comment:
How is the general view regarding external dependencies? An efficient implementation of the CheckTable would be available as external dependency at https://git.floragunn.com/util/fluent-collections (Apache 2 license). Would be a inclusion of that be fine?

Reply:
Pulling the library in as a maven dependency seems fine, I've got no issue with it.